sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -196,6 +212,142 @@ public virtual Response<UploadLogsResult> Upload<T>(string ruleId, string stream
+                            if (!runningTask.IsCompleted)
+                            {
+                                continue;
+                            }
+
+                            await runningTask.ConfigureAwait(false);
+                            runningTasks.RemoveAt(i);
+                            i--;
+                        }
+                    }
+                }
+
+                // Wait for all the remaining blocks to finish staging and then
+                // commit the block list to complete the upload
+                await Task.WhenAll(runningTasks).ConfigureAwait(false);
+
+                // Process all errors after tasks are done to determine status
+                // Will run on a single thread
+                foreach (Task<RunningTask<T>> task in runningTasks)
+                {
+                    // go through errors from each task and add to error list the response will be generated from
+                    foreach (UploadLogsError logsError in task.Result.ErrorList)
+                    {
+                        errors.Add(logsError);
+                    }
+                    // calculate the status using the helper method Status
+                    UploadLogsResult finalResult = new UploadLogsResult(errors, Status(logEntries, errors));
+                    return Response.FromValue(finalResult, response);
+                }
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+            finally
+            {
+                scope.Dispose();
+            }
+            return null;
+        }
+
+        private Task StageBatchAsync<T>(BatchedLogs<T> batch, string ruleId, string streamName, UploadLogsOptions options, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        private async Task<RunningTask<T>> CommitBatchListAsync<T>(BatchedLogs<T> batch, string ruleId, string streamName, CancellationToken cancellationToken)
+        {
+            List<UploadLogsError> errors = new();
+            RequestContext requestContext = GenerateRequestContext(cancellationToken);
+            Response response = null;
+
+            using HttpMessage message = CreateUploadRequest(ruleId, streamName, batch.LogsData, "gzip", requestContext);
+            response = await _pipeline.ProcessMessageAsync(message, requestContext, cancellationToken).ConfigureAwait(false);
+            if (response.Status != 204) // if any error is thrown log it
+            {
+                RequestFailedException requestFailedException = new RequestFailedException(response);
+                ResponseError responseError = new ResponseError(requestFailedException.ErrorCode, requestFailedException.Message);
+                List<Object> objectLogs = new List<Object>((IEnumerable<object>)batch.LogsList);
+                errors.Add(new UploadLogsError(responseError, objectLogs));
```

Comment:
Why is errors a list? There seems to be at most one error.

Reply:
Errors here represents the list of logs that have failed. It's always going to be one list of logs so I can update the name to represent that so it's less confusing

Reply:
but aren't we only adding one value here? I'm curious why `errors` is defined as a `List<UploadLogsError>`.